### ISSUE TITLE
Module execution order can be incorrect

### ISSUE THREAD

**Issue by evanw** [rollup/rollup#3888]
- Rollup Version: 2.34.0
- Operating System (or Browser): Chrome 86
- Node Version (if applicable): N/A
- Link to reproduction (IMPORTANT, read below): [link](http://rollupjs.org/repl/?version=2.34.0&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCcuJTJGbGliMiclNUNuaW1wb3J0JTIwJy4lMkZsaWIxJyU1Q25pZiUyMCh3aW5kb3cubGliJTIwISUzRCUzRCUyMCdsaWIxJyklMjB0aHJvdyUyMCdmYWlsJyUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1haW4yLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCcuJTJGbGliMSclNUNuaW1wb3J0JTIwJy4lMkZsaWIyJyU1Q25pZiUyMCh3aW5kb3cubGliJTIwISUzRCUzRCUyMCdsaWIyJyklMjB0aHJvdyUyMCdmYWlsJyUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmxpYjEuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyd2luZG93LmxpYiUyMCUzRCUyMCdsaWIxJyUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJsaWIyLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMndpbmRvdy5saWIlMjAlM0QlMjAnbGliMiclMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==)

I'm currently working on fixing some code splitting ordering issues in esbuild (https://github.com/evanw/esbuild/issues/399) and I noticed that Rollup has a similar problem. Here's an example problem case:

```js
// main.js
import './lib2'
import './lib1'
if (window.lib !== 'lib1') throw 'fail'
```

```js
// main2.js
import './lib1'
import './lib2'
if (window.lib !== 'lib2') throw 'fail'
```

```js
// lib1.js
window.lib = 'lib1'
```

```js
// lib2.js
window.lib = 'lib2'
```

I have a way to solve this in esbuild but it's a pretty big rewrite. Solving this is potentially a step up in complexity, so it's understandable why Rollup hasn't tackled this yet.

I figured I should log this in case it's not already a known issue. I did a quick search and the best I found was https://github.com/rollup/rollup/issues/3356, but it doesn't seem like this exact issue is covered by that one. My apologies if this issue is already known. If that's the case, feel free to close this.

### Expected Behavior

Both generated entry point chunks respect the correct module execution order and don't throw.

### Actual Behavior

One of the generated entry point chunks that Rollup generates will throw because common code that is shared between both of the entry points is evaluated in the wrong order. It's easy to see why this happens from the generated code:

```js
// output/main.js
import './lib1-4ba6e17e.js';
if (window.lib !== 'lib1') throw 'fail'
```

```js
// output/main2.js
import './lib1-4ba6e17e.js';
if (window.lib !== 'lib2') throw 'fail'
```

```js
// output/lib1-4ba6e17e.js
window.lib = 'lib2';
window.lib = 'lib1';
```

**Comment by lukastaegert** [comment]
Yes, this is a known issue. I have an idea to fix this which would involve calculating the execution order for each entry point separately and then splitting into smaller chunks so that the relative execution order inside each chunk is correct for all entry points, but unfortunately there has been little time to tackle this. It is no small thing and I am especially wary of the performance cost (which may not be that much of an issue for you).
Similar issues arise when there are circular dependencies that can start executing at different points depending on the entry point.

**Comment by evanw** [comment]
Yes, that's one solution I considered. But people have already been complaining about esbuild already generating multiple chunks and that approach would likely generate way more chunks. So that seems somewhat undesirable.

FWIW the other approach that I'm currently considering more strongly is to make code in shared chunks lazily-evaluated, at least in cases like this.

Say you have some shared code like this:

```js
export function fn() {}
export class Foo { ... }
export let foo = something()
```

You could transform it to something like this:

```js
export function fn() {}
export let Foo
export let foo

function __init() {
  Foo = class { ... }
  foo = something()
}
```

Then entry points could call the `__init` functions in the correct order for that entry point. Multiple entry points with different evaluation orders could then share code in a common chunk without evaluation order problems. Doing this is complex of course, but it seems like it should be able to emulate ESM semantics perfectly, even in non-ESM output formats. I'm currently working on a rewrite of esbuild to get it to work something like this.

> Yes, this is a known issue.

Glad to hear that it's already known. Feel free to close this issue then if it's already tracked elsewhere.

**Comment by guybedford** [comment]
It is worth carefully bearing in mind the underlying use cases in play here.

If you consider that entry points are routes of an application (`main1.js` and `main2.js` in the example), then a dynamic rendering of either route on the static instantiation of the other route is a perfectly fine thing to do in the application, and the assumption then is that the code is written such that the ordering differences between `lib1` and `lib2` do not apply.

Then for the actual use case of non-idempotent global side effects, the main thing it seems to apply to today is eg different versions of libraries that define a global. A use case which once every library is aways and only a module should go away in time.

For dynamic `import()` "Dynamic entry points", that are detected during the initial module graph trace, there does exist the concept of entry points dependence - if an entry point is only used as a dynamic import, and is not a top-level "public entry point" otherwise, then we know the full static graph of those modules that dynamically import it will always have an earlier ordering. This can in turn infer ordering assumptions and perhaps this kind of more involved ordering constraint handling is what @lukastaegert is referring.

As to the technique, that is a very interesting approach for handling execution differences, and perhaps there will be architectural knock-on effects to the codebase of this work such that this spearheading development of the codebase first and foremost.

I'd be weary of spending too much time on the use case if that is not the case though and I don't believe we will have plans to implement such techniques in RollupJS short of better ordering invariant handling as described by Lukas.

Thanks again for initiating discussion here, it's always great to have these cross-project discussions.

**Comment by ferferga** [comment]
@lukastaegert Hey, I know this kind of comments are useless, but it's been a really long time without any updates to this issue.

Is there any way we can help to implement the fix you had in mind?

Happy new year btw!
